### ISSUE TITLE
How to get 3D points in camera coordinate system

### ISSUE THREAD

**Issue by hahamini** [geopavlakos/hamer#60]
Thanks for sharing your awesome code.

To compute the pose of the hand, I need to get the 3D coordinates of each keypoint.

I ran demo.py with an image of a hand positioned approximately **60 to 70cm** away.
However, even though I added pred_cam_t_full as commented in #30 , I got the following result.

How can I get the 3D coordinates of the keypoints of the hand using the results predicted by the model?

And where in the code should I use the camera's focal length for "better hand position estimation" as you mentioned [#30 ]?

```
pred_cam_t_full[0]
-> array([ 0.13043386,  0.04035436, 18.107185  ], dtype=float32)

np.add(out['pred_vertices'][0][0].cpu().detach().numpy(), pred_cam_t_full[0])
-> array([2.5597063e-01, 6.0645305e-03, 1.8093334e+01], dtype=float32)
 
np.add(out['pred_keypoints_3d'][0][0].cpu().detach().numpy(), pred_cam_t_full[0])
-> array([ 0.22584924,  0.04672063, 18.113363  ], dtype=float32)
```

```
out['pred_keypoints_3d']
tensor([[ 0.0954,  0.0064,  0.0062],
        [ 0.1290, -0.0092, -0.0042],
        [ 0.1529, -0.0283, -0.0071],
        [ 0.1693, -0.0494, -0.0112],
        [ 0.1883, -0.0758, -0.0247],
        [ 0.1343, -0.0735, -0.0098],
        [ 0.1391, -0.1052, -0.0175],
        [ 0.1431, -0.1253, -0.0259],
        [ 0.1458, -0.1499, -0.0322],
        [ 0.1123, -0.0838, -0.0160],
        [ 0.1099, -0.1141, -0.0248],
        [ 0.1075, -0.1367, -0.0298],
        [ 0.1037, -0.1626, -0.0306],
        [ 0.0889, -0.0738, -0.0236],
        [ 0.0911, -0.1021, -0.0283],
        [ 0.0904, -0.1264, -0.0334],
        [ 0.0895, -0.1519, -0.0369],
        [ 0.0727, -0.0621, -0.0319],
        [ 0.0681, -0.0826, -0.0334],
        [ 0.0625, -0.1006, -0.0357],
        [ 0.0580, -0.1211, -0.0405]], device='cuda:0')
```
```
out['pred_vertices']
tensor([[ 0.1255, -0.0343, -0.0139],
        [ 0.1234, -0.0240, -0.0139],
        [ 0.1328, -0.0301, -0.0156],
        ...,
        [ 0.0857, -0.0462, -0.0344],
        [ 0.0860, -0.0568, -0.0347],
        [ 0.0920, -0.0293, -0.0262]], device='cuda:0')
```

**Comment by geopavlakos** [comment]
The translation of the hand in the camera frame is computed [here](https://github.com/geopavlakos/hamer/blob/dc19e5686198a7c3fc3938bff3951f238a85fd11/demo.py#L144). The variable `scaled_focal_length` corresponds to the focal length that is used to compute this translation. You might want to change that value with the focal length of your camera.
Please check [this issue](https://github.com/geopavlakos/hamer/issues/9#issuecomment-1865148920) for an explanation.

**Comment by hahamini** [comment]
@geopavlakos Thanks.
I did a calculation to compare the predicted value with the actual value. It showed this much error. What is the reason? According to your paper, the 3D pose accuracy is less than 1cm.

Results,
```
real:  [[ 0.05832911 -0.11363396  0.806     ]]
pred:  [[ 0.05534421 -0.10781892  0.76475427]]
```

```python
from live import load_for_test

data, img_cv2, depth = load_for_test('test') # Data from model's out, pred_cam_t_full from demo.py

image = img_cv2.astype(np.float32)[:, :, ::-1] / 255.0
vertices = data['vertices']
pred_cam_t_full = data['pred_cam_t_full']
cam_t = data['cam_t']
camera_matrix = data['cam_matrix']
crop_img = ((data['crop_img_np'] * np.array([0.229, 0.224, 0.225]).reshape(3,1,1)) + np.array([0.485, 0.456, 0.406]).reshape(3,1,1)).transpose((1,2,0))
is_right = np.array([1], dtype=np.float32)

verts = vertices
verts[:, 0] = (2 * is_right -1) *verts[:, 0]

image_size = img_cv2.shape[:2][::-1]

print(verts.shape)
verts = verts[:,0,:] # select 0 index
# index_finger_tip = vertices[320]
# thumb_tip = vertices[744]

print('camera_matrix', camera_matrix)
print('camera_translation', cam_t)
print('pred_cam_t_full', pred_cam_t_full)
print('cam_t', cam_t)
print('img_size', image_size)

def get_xyzs_from_depth(
    depth: np.ndarray,
    xys: np.ndarray,
    intrinsic,
    radius: int = 1,
    depth_scale: float = 0.001,
):
    fx = intrinsic[0, 0]
    fy = intrinsic[1, 1]
    cx = intrinsic[0, 2]
    cy = intrinsic[1, 2]

    if len(xys.shape) != 2:
        xys = np.array([xys])

    xs, ys = xys[:, 0], xys[:, 1]
    if radius <= 1:
        zs = depth[ys.astype(int), xs.astype(int)] * depth_scale
    # else:
    #     zs = circle_average(depth, xys, radius, np.median) * depth_scale

    xs = (xs - cx) / fx * zs
    ys = (ys - cy) / fy * zs

    indices = np.where(zs != 0.0)

    return np.array([xs[indices], ys[indices], zs[indices]]).T

def project_to_2d(vertices, K):
    vertices_2d = []
    for vertex in vertices:
        projected = K @ (vertex + pred_cam_t_full).T
        projected = projected / projected[-1]
        vertices_2d.append(projected[:2])
    return np.array(vertices_2d)

vertices_2d_1 = project_to_2d(verts, camera_matrix).reshape(-1, 2)
print(vertices_2d_1)
print('real: ', get_xyzs_from_depth(depth, vertices_2d_1, camera_matrix))
print('pred: ', np.add(verts, pred_cam_t_full))
```

output
``` 
(1, 778, 3)
camera_matrix [[907.41577148   0.         643.01605225]
 [  0.         907.24542236 364.14193726]
 [  0.           0.           1.        ]]
camera_translation [[-0.13099778 -0.07278255  5.08229494]]
pred_cam_t_full [[-0.07502446 -0.14791694  0.78469563]]
cam_t [[-0.13099778 -0.07278255  5.08229494]]
img_size (1280, 720)
[[708.68447666 236.23389071]]
real:  [[ 0.05832911 -0.11363396  0.806     ]]
pred:  [[ 0.05534421 -0.10781892  0.76475427]]
```

**Comment by geopavlakos** [comment]
The errors reported in the paper are typically measuring the _hand pose_ accuracy and are related to the 3D pose of the fingers with respect to the wrist. The error you are measuring is related to the _absolute 3D location_ of the hand, which can be less accurate due to the scale ambiguity in monocular data.

I believe that the ~4cm error in the absolute 3D location accuracy sounds reasonable, although you might want to make sure that the vertex you select is indeed visible in the frame you consider. Otherwise the measurement from the depth sensor would be less accurate.
